### PR TITLE
Add additional methods for supporting search for bundles, collections, observationals, processing levels, instruments, and investigations

### LINKED ISSUE #29: As a user, I want to search for products based upon a target LID
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data User

### 💪 Motivation

...so that I can find products associated with a specific target

### 📖 Additional Details

_No response_

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

_No response_

### 🎉 I&T

_No response_

### LINKED ISSUE #30: As a user, I want to search for products based upon a investigation LID
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data User

### 💪 Motivation

...so that I can find products associated with a specific investigation

### 📖 Additional Details

_No response_

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

_No response_

### 🎉 I&T

_No response_

### LINKED ISSUE #31: As a user, I want to search for products based upon a instrument LID
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data User

### 💪 Motivation

...so that I can search for products of a specific instrument

### 📖 Additional Details

_No response_

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

_No response_

### 🎉 I&T

_No response_

### LINKED ISSUE #32: As a user, I want to search for observational products only
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data User

### 💪 Motivation

...so that I can filter my search to just observational products

### 📖 Additional Details

_No response_

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

_No response_

### 🎉 I&T

_No response_

### LINKED ISSUE #33: As a user, I want to search for collection products only
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data User

### 💪 Motivation

...so that I can get only collection products

### 📖 Additional Details

_No response_

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

_No response_

### 🎉 I&T

_No response_

### LINKED ISSUE #34: As a user, I want to search for bundle products only
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data User

### 💪 Motivation

...so that I can filter on only bundle products

### 📖 Additional Details

_No response_

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

_No response_

### 🎉 I&T

_No response_

### LINKED ISSUE #35: As a user, I want to filter on a specific processing level
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data User

### 💪 Motivation

...so that I can filter on a specific processing level

https://pds.nasa.gov/datastandards/documents/im/v1/index_1M00.html#attribute_pds_primary_result_summary_pds_processing_level

### 📖 Additional Details

_No response_

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

_No response_

### 🎉 I&T

_No response_

### PR DESCRIPTION
**Pull request by jordanpadams**
## 🗒️ Summary
Add the following methods:
* `has_processing_level`
* `has_instrument`
* `bundles`
* `collections`
* `observationals`
* `has_investigation`

## ⚙️ Test Data and/or Report
Yeah... I tested it locally? But we probably need to figure out a good way to develop integration tests for this.

## ♻️ Related Issues

resolves #29
resolves #30
resolves #31
resolves #32
resolves #33
resolves #34
resolves #35

### CODE REVIEW

**Review by nutjob4life — APPROVED** [review]
Assuming `__add_clause` works, this looks fine 👍

**Comment by jordanpadams** [comment]
@tloubrieu-jpl would you mind taking a look when you have a chance? you have a vision for the architecture here, so I want to make sure this is inline with what you were thinking.

**Comment by tloubrieu-jpl** [comment]
@jordanpadams sorry I missed to look at it earlier. I will try to review it update when I have chance. This looks good but I have a couple of things I would like to look at:
- get: there might be a more efficient way to implement the query, not using the q parameters
- a few inline help are copy paste and need updates

**Comment by tloubrieu-jpl** [comment]
I also would like to have test for each implemented method.

**Comment by jordanpadams** [comment]
Mistakenly closed without merging

File: src/pds/updart/client.py [hunk 1]
```diff
@@ -94,6 +110,85 @@ def of_collection(self, id: str):
...
+        """Selects products with specific processing level.
+
+        :param processing_level: telemetry, raw, partially-processed, calibrated, derived
+        :return: a Products instance with processing level = Derived filter applied
+        """
+        clause = f'pds:Primary_Result_Summary.pds:processing_level eq "{processing_level.title()}"'
+        self.__add_clause(clause)
+        return self
+
+    def get(self, id: str):
```
Comment (tloubrieu-jpl) [inline src/pds/updart/client.py:174]: inline help is a copy/paste, need to be revised

File: src/pds/updart/client.py [hunk 2]
```diff
@@ -94,6 +110,85 @@ def of_collection(self, id: str):
...
+    def has_instrument_host(self, id: str):
+        """Selects products that have instrument host.
+
+        :return: a Products instance with Product_Observational class filter applied
+        """
+        clause = f'ref_lid_instrument_host eq "{id}"'
+        self.__add_clause(clause)
+        return self
+
+    def has_processing_level(self, processing_level: PROCESSING_LEVELS = "raw"):
```
Comment (tloubrieu-jpl) [inline src/pds/updart/client.py:164]: test need to be added

File: src/pds/updart/client.py [hunk 3]
```diff
@@ -94,6 +110,85 @@ def of_collection(self, id: str):
...
+    def has_instrument(self, id: str):
+        """Selects products that have instrument.
+
+        :return: a Products instance with Product_Observational class filter applied
+        """
+        clause = f'ref_lid_instrument eq "{id}"'
+        self.__add_clause(clause)
+        return self
+
+    def has_instrument_host(self, id: str):
```
Comment (tloubrieu-jpl) [inline src/pds/updart/client.py:155]: test need to be added

File: src/pds/updart/client.py [hunk 4]
```diff
@@ -94,6 +110,85 @@ def of_collection(self, id: str):
...
+        clause = 'product_class eq "Product_Collection"'
+        self.__add_clause(clause)
+
+        if type:
+            clause = f'pds:Collection.pds:collection_type eq "{type}"'
+            self.__add_clause(clause)
+
+        return self
+
+    def bundles(self):
```
Comment (tloubrieu-jpl) [inline src/pds/updart/client.py:137]: test need to be added

File: src/pds/updart/client.py [hunk 5]
```diff
@@ -94,6 +110,85 @@ def of_collection(self, id: str):
...
+    def observationals(self):
+        """Selects observational products for a specific filter.
+
+        :return: a Products instance with observational product class filter applied
+        """
+        clause = 'product_class eq "Product_Observational"'
+        self.__add_clause(clause)
+        return self
+
+    def collections(self, type: Optional[str] = None):
```
Comment (tloubrieu-jpl) [inline src/pds/updart/client.py:122]: test need to be added

File: src/pds/updart/client.py [hunk 6]
```diff
@@ -62,6 +66,18 @@ def has_target(self, identifier: str):
         self.__add_clause(clause)
         return self
 
+    def has_investigation(self, identifier: str):
```
Comment (tloubrieu-jpl) [inline src/pds/updart/client.py:69]: test need to be added

File: src/pds/updart/client.py [hunk 7]
```diff
@@ -94,6 +110,85 @@ def of_collection(self, id: str):
...
+    def get(self, id: str):
+        """Selects products which belong to a collection.
+
+        :param id: product identifier
+        :return: a Products instance with identifier applied
+        """
+        self.__add_clause(f'lidvid like "{id}"')
+        return self
+
+    def filter(self, clause: str):
```
Comment (tloubrieu-jpl) [inline src/pds/updart/client.py:183]: test need to be added.

docstring copy/paste need to be updated

**Review by tloubrieu-jpl — CHANGES_REQUESTED** [review]

File: src/pds/updart/client.py [hunk 8]
```diff
@@ -62,39 +66,138 @@ def has_target(self, identifier: str):
...
+        """Selects Observational products for a specific filter.
+
+        :return: a Products instance with Observational product class filter applied
+        """
+        clause = 'product_class eq "Product_Observational"'
+        self.__add_clause(clause)
+        return self
+
+    def collections(self, type: Optional[str] = None):
+        """Selects Collection products for a specific filter.
```
Comment (tloubrieu-jpl) [inline src/pds/updart/client.py:123]: Not a critical feedback; I don't understand what the meaning of "for a specific filter" in this context.
I would say "Selects products of class Collection only" or "Select Collections only".

File: src/pds/updart/client.py [hunk 9]
```diff
@@ -62,39 +66,138 @@ def has_target(self, identifier: str):
...
+        self.__add_clause(clause)
+
+        if type:
+            clause = f'pds:Collection.pds:collection_type eq "{type}"'
+            self.__add_clause(clause)
+
+        return self
+
+    def bundles(self):
+        """Selects Bundle products for a specific filter.
```
Comment (tloubrieu-jpl) [inline src/pds/updart/client.py:138]: Same as above

**Review by tloubrieu-jpl — APPROVED** [review]
That looks good, thanks @collinss-jpl .
I have minor comments that I shared here while I was going through the PR, but it is approved.

**Pull request merged**